hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AssociateTokenRecipientsStep.java

```diff
@@ -184,62 +197,27 @@ private TokenAssociation validateAndBuildAutoAssociation(
+                } catch (IllegalArgumentException ignore) {
+                    throw new HandleException(INSUFFICIENT_PAYER_BALANCE);
+                }
             }
+            final var newRelation = autoAssociate(account, token, accountStore, tokenRelStore, config);
+            return asTokenAssociation(newRelation.tokenId(), newRelation.accountId());
         } else {
             validateTrue(tokenRel != null, TOKEN_NOT_ASSOCIATED_TO_ACCOUNT);
             validateFalse(tokenRel.frozen(), ACCOUNT_FROZEN_FOR_TOKEN);
             return null;
         }
     }
 
-    /**
-     * Dispatches a synthetic transaction to associate the token with the account. It will increment the usedAutoAssociations
-     * count on the account and set the token relation as automaticAssociation.
-     * This is done only if the unlimitedAutoAssociationsEnabled is enabled.
-     * @param token The token to associate with the account
-     * @param accountStore The account store
-     * @param tokenRelStore The token relation store
-     * @param context The context
-     * @param account The account to associate the token with
-     */
-    private void dispatchAutoAssociation(
-            final @NonNull Token token,
-            final @NonNull WritableAccountStore accountStore,
-            final @NonNull WritableTokenRelationStore tokenRelStore,
-            final @NonNull HandleContext context,
-            final @NonNull Account account) {
-        final var accountId = account.accountIdOrThrow();
-        final var tokenId = token.tokenIdOrThrow();
-        final var syntheticAssociation = TransactionBody.newBuilder()
-                .tokenAssociate(TokenAssociateTransactionBody.newBuilder()
-                        .account(account.accountId())
-                        .tokens(token.tokenId())
-                        .build())
-                .build();
-        // We don't need to verify signatures for this internal dispatch. So we specify the keyVerifier to null
-        context.dispatchRemovablePrecedingTransaction(
-                syntheticAssociation, SingleTransactionRecordBuilder.class, null, context.payer());
-        // increment the usedAutoAssociations count
-        final var accountModified = requireNonNull(accountStore.getAliasedAccountById(accountId))
-                .copyBuilder()
-                .usedAutoAssociations(account.usedAutoAssociations() + 1)
-                .build();
-        accountStore.put(accountModified);
-        // We need to set this as auto-association
-        final var newTokenRel = requireNonNull(tokenRelStore.get(accountId, tokenId))
-                .copyBuilder()
-                .automaticAssociation(true)
-                .build();
-        tokenRelStore.put(newTokenRel);
+    private long associationFeeFor(@NonNull final HandleContext context, @NonNull final TransactionBody txnBody) {
+        final var topLevelPayer = context.payer();
+        // If we get here the payer must exist
```

Comment:
What does this comment mean?  `HandleContext.payer()` is declared to return `@NonNull` and is simply a getter.  So it isn't referring to L217 immediately above.

Should this be a comment on the method itself (like documenting a precondition)?

Reply:
Fixed.